### DISCUSSION TITLE
Implementation of hardware RNGs on MCUs

### DISCUSSION THREAD

**Discussion by MicroControleurMonde** [micropython/micropython#16420] [Hardware & Peripherals] [answered]
Hi guys,

I needed to generate random numbers several weeks ago and so I pulled out my RP2040 to do this. And then you know what it is, from thread to needle, I jumped on my esp32 and so on.

In fact I realized that the hardware RNG part wasn't there, so I started working on it.  I have been rummaging through the docs and I have coded the calls to the correct registers.

Question, I had fun on the PyBoard with **'pyb.rng()'**. Does it call on the STM32 hardware RNG  ? I didn't go to check.

[ESP32](https://github.com/MicroControleurMonde/ESP32_RNG)
[ESP32-C3](https://github.com/MicroControleurMonde/ESP32-C3_RNG)
[ESP32-C6](https://github.com/MicroControleurMonde/ESP32-C6_RNG)
[RP2040](https://github.com/MicroControleurMonde/RP2040-RNG) using ADC.
[Teensy 4.1](https://github.com/MicroControleurMonde/Teensy_4.1_RNG)

Note: For the Tensy, I am not able to implement the TRNG ... I used several ADC to generate values instead.

Cheers.

**Comment by robert-hh** [comment]
All three ports you mention support already a hardware TRNG with the `os.urandom()` method. The `random` class uses a Pseudo RNG.
See https://docs.micropython.org/en/latest/library/os.html#os.urandom and https://docs.micropython.org/en/latest/library/random.html#module-random

**Comment by MicroControleurMonde** [comment]
Hi Robert,

 Pseudo RNG ... **sorry** but If I have read the ESP32 docs correctly, they are hardware RNG, not pseudo-random.
 So what I've coded is going to read hardware registers.
 For the RP2040 and the Teensy, the ADCs are used to collect noise.
 I didn't touch the Pyboard. I used it as it is. And it's good 
 
 Now for **`urandom`**: "*Whenever possible, it is generated by the hardware random number generator*"
 Ok, good. 
 Which MCUs exactly ?

**Comment by robert-hh** [comment, marked as answer]
As said, the MicroPython module uses a Pseudo RNG in all ports.
`os.urandom()` is based on a True RNG (hardware), but not supported by all ports and MCUs.
Ports that uses a TRNG with os.urandom():

- STM32
- ESP32
- ESP8266
- RP2
- MIMXRT (teensy)
- SAMD
- Renesas RA
- NRF

Maybe more. In order to tell, whether a certain MCU uses a TRNG, look whether `os.urandom()` exists.

**Reply by MicroControleurMonde** [reply]
Thanks for the reply Robert.

I was wondering since the definition of **`urandom()`** was still rather vague.

![image](https://github.com/user-attachments/assets/b597c79d-2130-4571-81fb-066a92b90669)

You can admit that this is already confusing, right?

I'm delighted to discover that the i.MX RT are supported. Given the number of registers at stake, this will save me a lot of headaches.

Incredible: ‘RP2350 **contains** a Real Random Number Generator block based on Arm IP’.  They never cease to impress me at Raspberry!

**Reply by robert-hh** [reply]
Yes, I know this sentence.

**Comment by MicroControleurMonde** [comment]
- ESP8266 since when ?

![image](https://github.com/user-attachments/assets/b95a1cb8-b01d-4dd1-9b85-a0e40b77e241)

You're sure about that ?

**Reply by robert-hh** [reply]
As far as I understand the code, the random numbers are taken from the WiFi block as entropy source. Similar like you wanted it to do with ADC. Only in the WiFi section the bandwidth is much higher. I faintly remember that the ESP32 (maby only some of them) also use the WiFi block as entropy source,

The RP2040 also supports Hardware RNG, and it seems to use a side effect of the clock circuity.
For SAMD21 I used a jitter between two PLLs as random source.

I did not look into the other devices. And for each port there may be MCUs that do not support a TRNG.

**Reply by Josverl** [reply]
Would it make sense to:
 
- [ ] document the known implemented TRNGs in a Wiki page 
- [ ] add a link to that from the `random.urandom()` documentation

**Reply by MicroControleurMonde** [reply]
@robert-hh 

Not entirely: the ESP32s have a dedicated block for random generation.

In the Espressif doc, they advise this:

*"...Thermal noise comes from the high-speed ADC or SAR ADC or both"* and then ...*"When there is noise coming from the SAR ADC, the random number generator is fed with a 2-bit entropy in on clock cycle of RC_FAST_CLK (8 MHz), **which is generated from an internal RC oscillator** ...*".
Then you go read your register to collect your 32 bits value.

So, if my understand is correct, the term ‘RC oscillator’ refers to an internal oscillator based on an RC (resistor-capacitor) circuit. In the context of the ESP32 and its  RNG, this means that the random noise used to generate random numbers comes from the oscillator. Not only the ADC (or SAR ADC). They are using a XOR on the 3 sources.

From what I found in the datasheets, here are the boards which have an included  hardware RNG:  all ESP32, some STM32 (_STM32F4,STM32L4_), some Microchip SAMD (_SAMD21, SAMD51)_, surprisingly the BBC V2.2 (_Nordic nRF52833_) and the NXP  i.MX RT serie. And now the RP2350 ^_^

- The RP2040 does not have a built-in hardware random number generator. And indeed I used the ADC to replicate the ESP principle. 
It works but the final entropy is a bit lower ... 
- I did the same with the ESP8266 (ADC reading **+ Wifi**) and the entropy is a bit better than the RP2040.

**Reply by MicroControleurMonde** [reply]
@Josverl 

Well, that's totally make sens, because if I'd known about the **`urandom()`** module, I wouldn't have gone off fiddling with registers and I'd have focused on my entropy study.

**Reply by robert-hh** [reply]
The RP2040's data sheet section 2.17.5 is called `2.17.5. Random Number Generator`. It refers to `RANDOMBIT` bit in the ROSC register, which accordingly is used by machine.urandom() to get random numbers. I made some simple test for randomness, and it looked not very well. Even the simple histograms looked odd. But feeding the data stream though a CRC or even just XOR made it looked better. So it should pass RNG tests when used to constantly inject a Pseudo-RNG.

**Comment by MicroControleurMonde** [comment]
## And right you are !

Raspberry Pi Pico Datasheet -  Build-date: 2024-10-15 / Build-version: eec2b0c-clean

### Chapter 2.17. Ring Oscillator (ROSC) - Page 222.

![image](https://github.com/user-attachments/assets/b75cebb4-8b82-485e-a09a-b7a9b5501b36)

My datasheet was stamped Build-date **2023-06-14** Build-version a6fe703-clean ...
### My bad !
:-(

The ROSC is the internal oscillator that can be used to generate random numbers indeed.
Ok then, I will update my **RP2040_RNG** repo and test a fat batch of 1 million+ values to see if it pass the Dieharder and what entropie score it reaches.

In initial method (with ADC), it reached a **3.437** bits per byte entropy.
I will keep you posted.

**Reply by MicroControleurMonde** [reply]
OK I just finished to change my code (with use of RSOC).
One million generated.
And ... 
**Entropy** = **3.450903** bits per byte.
Slightly better.

Il will push the code today.

**Reply by MicroControleurMonde** [reply]
OK.

Estimated time to generate **1'000'000** values: **1400 seconds** (approx. 23 minutes)
Average time per value: **0.0014 seconds**
Estimated Throughput: ≈ **2857 Bytes/sec**
Number of values generated per second: ≈ **714 values/sec**

Code, tests and reports here:
https://github.com/MicroControleurMonde/RP2040_ROSC_RNG

**Reply by scruss** [reply]
> **Entropy** = **3.450903** bits per byte.

Random 32-bit integers printed as ASCII strings are around 3.45 bits of entropy per byte (since there are 11 possible characters in the output, and log₂ 11 ≅ 3.459). Are you sure you're checking the correct output?

Binary file output from `os.urandom()` generated on a Raspberry Pi Pico gives me 7.9998 bits of entropy per byte, according to `ent`. If I convert the output to decimal strings, I get 3.451 bits of entropy per byte.

**Reply by robert-hh** [reply]
@scruss Did you use a RP2040 or RP2350 for you test?

**Reply by robert-hh** [reply]
There is a difference between reading the ROSC RANDOMBIT and calling os.urandom(). os.urandom() uses ROSC RANDOMBIT as well, but feeds it through a CRC. See main.c. Therefore the random numbers look better and have a clean frequency distribution and can pass RNG test. But that does not increase the entropy of the source.
This feeding through a CRC was suggested by me in March 32. Some histograms taken at that time of 100kB data sets.
I did not test the data with e.g. the Diehard test suite.

Raw RANDOMBIT, read fast

![hist_rng_raw_fast](https://github.com/user-attachments/assets/ec5e63db-c665-495a-893a-6ab2fc5b4e60)

Raw RANDOMBIT, read slow

![hist_rng_raw_slow](https://github.com/user-attachments/assets/be263c6c-7a9d-4bc7-9090-c4bffd517db0)

RANDOMBIT processed by the CRC_

![hist_rng_crc](https://github.com/user-attachments/assets/792d9bf9-e93a-4576-a2a8-4ed4c71be794)

**Reply by MicroControleurMonde** [reply]
> There is a difference between reading the ROSC RANDOMBIT and calling os.urandom(). os.urandom() uses ROSC RANDOMBIT as well, but feeds it through a CRC. See main.c. Therefore the random numbers look better and have a clean frequency distribution and can pass RNG test. But that does not increase the entropy of the source. This feeding through a CRC was suggested by me in March 32. Some histograms taken at that time of 100kB data sets. I did not test the data with e.g. the Diehard test suite.
> 
> 

Well done on the CRC trick!
Fast too.

On my initial implementation, I do a random generation + SHA2-256 (whitening). Same thing, it gives a better appearance but it won't change the source entropy anyway. Physics is physics. 

In any case, SHA2 is out of date, and NIST has been recommending SHA3 for cryptographic functions since 2015, I believe? FIPS-202.
I've started a SHA3-256 on the RP2040, but for the moment I've only finalised the Keccak-p[1600] permutation part (my head hurts)...
Here: [Keccak-p1600](https://github.com/MicroControleurMonde/SHA3/blob/main/keccakp1600.py)

Diehard, it's perfect but they recommend spitting out 13 million values to make a ‘good’ valid test. On a bare RP2040 or ESP32 ... it becomes a bit fastidious. Fortunately, my RP2040 Waveshare has been upgraded to 16 MB of flash. 

 ... March 32 ?!? What date is that? 2022 ?

I haven't done any graphics yet. I'll work on it later. I'd like to make a nice scatter representation of entropy. I've found a good code to do it.

**Reply by MicroControleurMonde** [reply]
> > **Entropy** = **3.450903** bits per byte.
> 
> Random 32-bit integers printed as ASCII strings are around 3.45 bits of entropy per byte (since there are 11 possible characters in the output, and log₂ 11 ≅ 3.459). Are you sure you're checking the correct output?
> 
> Binary file output from `os.urandom()` generated on a Raspberry Pi Pico gives me 7.9998 bits of entropy per byte, according to `ent`. If I convert the output to decimal strings, I get 3.451 bits of entropy per byte.

Hi Sruss
**"Are you sure you are checking the correct output?"**
Well I thought so yes. But now I'm already starting to wonder... I think I will quickly generate  '.bin'  files and review all my calculations...

Thank you for the post, I was sincerely starting to despair of not passing an entropy of 4.
**7.9998** it's really excellent

**Reply by scruss** [reply]
> @scruss Did you use a RP2040 or RP2350 for you test?

“… generated on a Raspberry Pi Pico” would indicate that yes, I did.

**Reply by MicroControleurMonde** [reply]
@scruss 

Fresh update.
I changed my code to binary output (on RP2040 using **only** the ROSC) and it's an entire new reality !

For a _small and quick_ 10'000 random values sample, Entropy tests:

- Entropy = **0.996693** bits per bit.
   7.973544 bits/byte.
   Entropy close to 1 means that the bits are almost equally distributed between 0 and 1.

- Optimum compression would reduce the size of this 320000 bit file by **0 percent.**
There is no significant redundancy in the data

- Chi square distribution for 320000 samples is 1465.84, and randomly would exceed this value less than 0.01 percent of the times.
 So, the data are statistically very close to a random distribution
 
 - Arithmetic mean value of data bits is 0.4662 (0.5 = random).
 
 - Monte Carlo value for Pi is 3.326732673 (error 5.89 percent).
 Still acceptable error for a calculation based on this method
 
 - Serial correlation coefficient is 0.001526 (totally uncorrelated = 0.0).
There is hardly any correlation between consecutive bits, which is expected for random sequence. The data is well mixed and unstructured (no repeating patterns).

We could always add a CRC or SHA layer on top of that, but I doubt that will ultimately be necessary for the moment, given the results.
I will do later.

Well, what a surprise !
I didn't think our brave little Pico was capable of such performances. I'm delighted.
So I need to regenerate all my value batches and redo all the tests.
Fair enough.

**Reply by MicroControleurMonde** [reply]
OK, updated performance test, Ent and Dieharder:
https://github.com/MicroControleurMonde/RP2040_ROSC_RNG

**Reply by MicroControleurMonde** [reply]
![image](https://github.com/user-attachments/assets/8bd641d3-4a03-4bfb-8c11-5cf501255149)
![image](https://github.com/user-attachments/assets/9d734bda-1ff7-4bc8-be2b-d701123753aa)

**Reply by MicroControleurMonde** [reply]
## For the records: RNG with ROSC reading + XOR/LCG + SHA2-256
Sample 10'000 values

**Ent test:**
- Value Char Occurrences Fraction
      0              1279184   0.499681
      1              1280816   0.500319
      Total:       2560000   1.000000
- **Entropy = 1.000000 bits per bit**.
- Optimum compression would reduce the size of this 2560000 bit file by 0 percent.
- Chi square distribution for 2560000 samples is 1.04, and randomly would exceed this value 30.77 percent of the times.
- Arithmetic mean value of data bits is 0.5003 (0.5 = random).
- Monte Carlo value for Pi is 3.147169670 (error 0.18 percent).
- Serial correlation coefficient is -0.001185 (totally uncorrelated = 0.0).

![image](https://github.com/user-attachments/assets/ec286e60-f171-42a5-be88-10e0d4e0d9dc)

![image](https://github.com/user-attachments/assets/cb670a4c-c0a3-48f0-bcc2-ca2fb68f1751)

**Reply by scruss** [reply]
Good to hear.

It looks like you're still getting the "combing" effect that [robert hh](https://github.com/orgs/micropython/discussions/16420#discussioncomment-11638230) saw in his "Raw RANDOMBIT, read slow" graph. I'd suggest that's showing a pattern in your data. Since the randombit + CRC used in os.urandom() is generating noise at over 120 KB/s on a Raspberry Pi Pico at 125 MHz, it seems as if it's doing quite a fine job already.

I haven't run os.urandom() output through rngtest or dieharder yet. I don't need a cryptographic grade RNG for what I do.

**Reply by MicroControleurMonde** [reply]
@scruss 

Hi Stewart,

Yes indeed, their is an obvious pattern in the data !
In fact either you read raw ROSC and get a 0.99+ entropy with a slight pattern ... or you do the other way and the entropy is perfect at 1 but the Dieharder shows more bigger patterns !

My next task is to test **[Raw Rosc** versus **Raw Rosc + PCG32]** and then, versus **[Raw Rosc + PCG32 + SHA2-256]** to see what it changes in terms of distrib./dispersion.

I will go for **rngtest** and **TestU01** as well.

Now, practically speaking (in the normies world), Mister John Blue doesn't need all this. Not everybody wants to rack their brains.
And that's where [Robert](https://github.com/orgs/micropython/discussions/16420#discussioncomment-11638230) solution is "elegant".
It is fast and efficient.

**Reply by MicroControleurMonde** [reply]
Update.

- **TestU01** (Rabbit battery): [Report](https://github.com/MicroControleurMonde/RP2040_ROSC_RNG/blob/main/Reports/TestU01_Rabbit.md) - " _...  the generator does not perfectly meet the criteria for statistical tests of random number quality._"
- **rngtest** : [Report](https://github.com/MicroControleurMonde/RP2040_ROSC_RNG/blob/main/Reports/rngtest_rp2040_rosc_rng_1425424.txt) -  **all the _FIPS tests failed_**. :eyes:

**Reply by MicroControleurMonde** [reply]
@robert-hh 

Your `RANDOMBIT` +`CRC` makes wonder !
`Ent `and `rngtest` are perfect.
Only Diehard shows some few weakness.

And it's fast:  1'5000'00 values: 133 seconds...

**Reply by robert-hh** [reply]
Thanks. I tried to make it efficient. CRC is always good for spreading bits.
Obviously using it for creating keys may not the best choice. But for other random number applications it works well.

**Reply by MicroControleurMonde** [reply]
It is (_efficient_) !

You could create keys with it but it would require post-RNG un-bias filter (ex.Von Neumann method).
PCG32 (from O'Neil) is very fast as well. It might replace CRC with a less "critical" value dispersion.

**Reply by MicroControleurMonde** [reply]
@robert-hh 

FYI: I tested the Von Neumann filter: it works well but **more than 50%** of the values go to the trash.

I did 2 batches test of 13 million without and with filter and the statistical differences are really significant. But at what cost.


If you wish I can share the tests reports.
- Ent/DJent
- rngtest
- dieharder
- booltest

**Comment by MicroControleurMonde** [comment]
I have prepared a paper on RNGs and their application. Well, it's in French at the moment... I still haven't finished translating it completely. I will publish it later this week on my page.

I am also working on a small high entropy TRNG project (MCU + amp. + capacitors + resistors). 
~~I hope I can break the 7+ bits per byte threshold.~~

**Comment by MicroControleurMonde** [comment]
@robert-hh 

Guten Morgen Robert,

I'm not going to open a new thread because the topic is related to RNGs.

I am implementing my RNG code on **ESP32-S3-DevKitC-1 v 1.1.**

By reading the Technical Reference Manual version 1.6 for ESP32-S3, page 921, the memory address of the register 'RNG_DATA_REG' is indicated as being: **0x6003_3110**

![image](https://github.com/user-attachments/assets/5584f11a-29a4-4c38-a67b-dc69b09fb1e6)

I have tried everything possible, but the result was always nill. So I have decided to go fishing for information on the Espressif website, in their source code. I finally found what I was looking for here (within the **wdev_reg.h** include file) :

https://github.com/espressif/esp-idf/blob/cc9fb5bd5ea7d3ddb1a8c26e5bf0076b3e1352f3/components/soc/esp32s3/include/soc/wdev_reg.h#L4 

![image](https://github.com/user-attachments/assets/62eddedc-736c-455d-9872-fb70398d7292)

I have tried this address and now it works perfectly.

## So, question: **Can anyone here confirm or deny that the address `0x6003_3110` is incorrect and that the correct address is potentially `0x6003507C` ?**

![image](https://github.com/user-attachments/assets/d170ddf7-f831-493e-a4f2-dbc2cb5eadfc)

[primary_test_code_esp32-s3_hw_rng.md](https://github.com/user-attachments/files/18408437/primary_test_code_esp32-s3_hw_rng.md)

I have send a report to Espressif's CRM website, but not sure they will reply soon ...

**Reply by MicroControleurMonde** [reply]
Hi,

**FYI:** I just received confirmation from Espressif about the ESP32 RNG_DATA_REG:

![image](https://github.com/user-attachments/assets/1b819c50-9203-4bf6-9a11-fef25e34a253)

One month.
Not that bad.

Cheers

**Comment by tferrin** [comment]
Following up on this discussion specifically with respect to the RP2350's TRNG, section 12.12 of the RP2350 datasheet says it is compliant with the FIPS 140-2 and NIST SP 800-90B standards. So if you need this kind of guarantee for random number generation you might consider using this hardware feature. This topic on the Raspberry Pi forum includes MicroPython code for generating 192-bit random numbers using the chip's TRNG: https://forums.raspberrypi.com/viewtopic.php?t=390946

**Reply by scruss** [reply]
As I said in that thread:

> I'm seeing more than 7.999 bits/byte entropy from `os.urandom()` on RP2040 and RP2350, similar to what I get from /dev/urandom and /dev/random (and /dev/hwrng on a Raspberry Pi 5)

As ever: доверяй, но проверяй

**Reply by tferrin** [reply]
No argument from me that os.urandom() produces good random values. My point was about compliance to FIPS and NIST standards. If your application needs to adhere to one of these standards then ARM has gone to the trouble of securing a compliant implementation.
